Docmentation update

CSSFrancis opened the pull request:
---
name: Documentation Update
about: Fixes some of the bigger problems with pyxem's documentation.

---

**Checklist**
- [x] Updated CHANGELOG.md
- [x] Clean up inter sphinx mapping
- [x] Autodoc to make documentation instead of explicit
- [x] Use pydata format
- [x] Add to user guide
- [x] Add Examples to documentation
- [x] Better installation user guide
- [x] Checkout demos from pyxem/pyxem-demos
- [x] Link Checking
    - [x] Add Link check CI
    - [x] Make sure all links work
- [x] Marked as finished

**What does this PR do? Please describe and/or link to an open issue.**
See #955

CSSFrancis commented:
@hakonanes Quick question about pooch/notebooks and doc building etc.  

I want to move the Jupyter notebooks out of the pyxem repo and have them just be directly pulled from the pyxem-demos repo.

There seem like a couple of different options:

1. Use pooch in the conf.py file to pull the Jupyter notebooks from the examples repo (I think this isn't the best way because the checksum might not stay the same from version to version. 
2. Use git in the ??Make?? file to checkout the examples and put them in the right repo if they aren't already checked out
3. Just add to the Read the docs .ymal and don't care about local making of the example notebooks unless necessary in which case someone can always just check them out themselves.

hakonanes commented:
I agree that fetching the notebooks using git seems like the best option. Using Pooch seems like overkill.

Just had a quick look over the docs. They look nice overall! Some issues with the examples gallery. Otherwise things are inplace and can be improved upon over time.

CSSFrancis commented:
@hakonanes I think I figured out the notebook fetching to some degree using git. I noticed the issue with the examples gallery as well but haven't figured out what actually causes it. It seems like the file names are showing up when they shouldn't be. It might be because some of the examples are blank?

hakonanes commented on doc/user_guide/index.rst:
```diff
@@ -0,0 +1,33 @@
+    :hidden:
+
+    ../tutorials/index.rst
+    ../auto_examples/index.rst
```
I believe this should be

```suggestion
    ../examples/index.rst
```

as per the [same file](https://raw.githubusercontent.com/pyxem/kikuchipy/develop/doc/user/index.rst) in kikuchipy's docs.

CSSFrancis replied:
It should be correct (I think) the difference is:

https://github.com/pyxem/kikuchipy/blob/b7d626ce52c9c049fefec59c0d103e5fc646434b/doc/conf.py#L338-L347

vs.

https://github.com/CSSFrancis/pyxem/blob/f1612457a3965a503265e46db499ec85052662b3/doc/conf.py#L178-L184

I just set that folder to `auto_examples`.  That is how `hyperspy` labels it and I think it helps to show that the folder shouldn't be edited.

hakonanes replied:
I agree that "auto" is a clear indication that the files are automatically generated. The URL will include "auto_examples" and not "examples", though (https://hyperspy--3238.org.readthedocs.build/en/3238/auto_examples/index.html). I find this more verbose, but it isn't a big issue.

hakonanes replied:
And then I don't know what causes the weird gallery.

hakonanes commented:
The docs seem to shape up nicely, great job! Just give me ping when you want a review.

CSSFrancis commented:
@hakonanes This should be good now if you want to look it over an review!

I think in order to run the Jupyter notebooks for read the docs we need a little bit more memory.  I guess we can always include the outputs for the notebooks in the `pyxes-demos`.

hakonanes commented:
OK, this is a large PR! Could you recommend a best way to review it? In what order should I go through the files?

The test suite run time increased from [22 min 40 s](https://github.com/pyxem/pyxem/actions/runs/6486528287) to [45 min 26 s](https://github.com/pyxem/pyxem/actions/runs/6512062519) after the current additions in this PR. Perhaps a separate flag to pass to pytest when these tests should be run? It might be best just to not run these tests in GitHub CI.

CSSFrancis commented:
> OK, this is a large PR! Could you recommend a best way to review it? In what order should I go through the files?

@hakonanes Sorry I kind of got a little carried away :) 

There shouldn't be any changes to the actual code, but there were some pretty big problems with the documentation so it took a little bit to clean it all up

The changes are:

1. Checkout the Jupyter Notebooks using git instead of having them in the pyxem repo.  This affects the `demos/` folder and the `read_the_docs.ymal` as well as the new `doc.ymal` work flows. The new doc workflow copies hyperspy to check the links.
2. Clean up API reference:  This probably touches the most files. These are just cleaning up linking in the `signals/` folder as well as making sure the each `__init__` file has a description included. I also copied kikuchipy for the `docs/_templates`
3. Added Pooch for data hosting and added `Examples`.  These are weird things to go together but I in order to get the Examples to run I wanted an easier way to load the data so I added in the `data._data` and `data._registry` files copied from kikuchipy.  The Examples I had intentions of adding a bunch of stuff but I realized (too late) that this PR had grown too large. 

> The test suite run time increased from [22 min 40 s](https://github.com/pyxem/pyxem/actions/runs/6486528287) to [45 min 26 s](https://github.com/pyxem/pyxem/actions/runs/6512062519) after the current additions in this PR. Perhaps a separate flag to pass to pytest when these tests should be run? It might be best just to not run these tests in GitHub CI.

Yea that was my plan.

ericpre commented:
> Failing is related to testing pooch. I think that it takes (a long time) to download from zenodo and it seems like it is pretty highly variable depending on the load zenodo is experiencing. I might change some of the smaller files to GitHub links as that seems a fair bit quicker.

The pooch/zenodo issue is related to https://github.com/fatiando/pooch/pull/373.

hakonanes commented:
Thank you for pointing to this issue, @ericpre. I just tested downloading a 500 MB EBSD dataset from Zenodo via kikuchipy's `kikuchipy.data.si_wafer(allow_download=True)`. This worked just fine. I used Pooch 1.7.0. So, the issue doesn't seem to affect everyone using Pooch to download from Zenodo.

OK, @CSSFrancis. I'll review this within the weekend.

hakonanes commented:
That must be the reason. Thank you so much for checking this!

ericpre commented:
I get a different error when I try `kikuchipy.data.si_wafer`: https://github.com/pyxem/kikuchipy/issues/654... :(

hakonanes commented on pyxem/data/_data.py:
```diff
@@ -0,0 +1,311 @@
+import os
+
+# Create a downloader object
+marshall = pooch.create(
```
Marshall was the name of my sister's dog. You're free to use it. However, this is a great opportunity to name this particular pooch after a dog in *your* life!

CSSFrancis replied:
:) I like that!

hakonanes commented on pyxem/data/_data.py:
```diff
@@ -0,0 +1,311 @@
+    ----------
+    **kwargs
+        Keyword arguments passed to :func:`~hyperspy.api.load`.
+    Examples
```
Blank line, please. And in the other functions.

hakonanes commented on pyxem/data/_data.py:
```diff
@@ -0,0 +1,311 @@
+    base_url="",
+    version=version.replace(".dev", "+"),
+    version_dev="develop",
+    env="PYXEM_DATA_DIR",
```
I suggest to describe in the docs how the cache location can be set using this environment variable. We've done so in [kikuchipy's docs](https://kikuchipy.org/en/stable/reference/generated/kikuchipy.data.html).

hakonanes commented on pyxem/data/_data.py:
```diff
@@ -0,0 +1,311 @@
+    path=pooch.os_cache("pyxem"),
+    base_url="",
+    version=version.replace(".dev", "+"),
+    version_dev="develop",
```
```suggestion
    version_dev="main",
```

hakonanes commented on pyxem/data/_data.py:
```diff
@@ -0,0 +1,311 @@
+    --------
+    >>> import pyxem as pxm
+    >>> s = pxm.data.au_grating()
+    >>> s
```
The next line should print what the signal's string representation.

hakonanes commented on pyxem/data/_data.py:
```diff
@@ -0,0 +1,311 @@
+)
+
+
+def au_grating(allow_download=False, **kwargs):
```
We've chosen to describe the license of each dataset in the `kikuchipy.data` module. Might be something to consider for pyxem as well.

hakonanes replied:
Also, if datasets are hosted elsewhere than in the package, a link to the host can be added to docstring notes.

CSSFrancis replied:
To be honest I don't really know what the best way to handle the license for each dataset.  Some of them are mine but some are from different people contributing to the pyxes-demos.

CSSFrancis replied:
Done! (Well I should probably include the license in zenodo as well but this should be good here

hakonanes commented on pyxem/data/_data.py:
```diff
@@ -0,0 +1,311 @@
+    file_package_path: Path
+    file_cache_path: Path
+    expected_md5_hash: str = ""
+    collection_name: Optional[str] = None
```
No datasets are retreived from collections. I suggest to remove this part from the class to reduce complexity.

We use this in kikuchipy's data module when we want to download single files from a single Zenodo upload containing multiple files. For example when we want one EBSD dataset (.zip file) from a Zenodo upload containing ten datasets (as ten .zip files).

hakonanes commented on pyxem/signals/__init__.py:
```diff
@@ -93,3 +88,7 @@
 ]
+
+
+def __dir__():
+    return sorted(__all__)
```
This is only needed for lazy loading of modules, I think. Do we plan to add that to pyxem?

CSSFrancis replied:
Ahh that's on my list (eventually). I think I must have just copied it without knowing what it was for....

CSSFrancis replied:
We are going to leave that for a separate PR though :)

hakonanes replied:
Sounds good! I suggest to remove it here, then.

hakonanes commented on pyxem/signals/diffraction2d.py:
```diff
@@ -104,7 +112,7 @@ def apply_affine_transformation(
         ----------
         D : array or Signal2D of arrays
-            3x3 np.array (or Signal2D thereof) specifying the affine transform
+            3x3 :py: (or Signal2D thereof) specifying the affine transform
```
?

hakonanes commented on pyxem/signals/diffraction2d.py:
```diff
@@ -1185,8 +1194,8 @@ def template_match_with_binary_image(
         --------
-        template_match_disk
-        template_match_ring
+        :meth:`~pyxem.signals.DiffractionSignal2D.template_match_disk`
```
I don't think the Sphinx syntax is necessary. Just the full path, like `pyxem.signals.DiffractionSignal2D`.

CSSFrancis replied:
I can't seem to get links like that to work without it...

hakonanes replied:
That's weird. Works in kikuchipy: [source](https://github.com/pyxem/kikuchipy/blob/develop/kikuchipy/signals/ebsd.py#L611-L616) -> [docs](https://kikuchipy.org/en/stable/reference/generated/kikuchipy.signals.EBSD.remove_dynamic_background.html).

hakonanes commented on pyxem/signals/polar_diffraction2d.py:
```diff
@@ -138,9 +142,10 @@ def get_full_pearson_correlation(
             overwritten.
-
+        kwargs: dict
+            Any additional options for the :meth:`~hyperspy.signal.BaseSignal.map` function
```
```suggestion
            Any additional options for the :meth:`~hyperspy.signal.BaseSignal.map` function.

```

hakonanes commented on readthedocs.yaml:
```diff
@@ -4,21 +4,24 @@
-formats: all
 
 # Set the version of Python and other tools
+# make demos is a custom command to fetch the demo notebooks
```
```suggestion
# check-demos is a custom command to fetch the demo notebooks
```

hakonanes reviewed (changes requested):
This is a great step towards a complete documentation of pyxem! The framework is in this PR. The documentation can be made more complete in later PRs.

I have smaller suggestions and a few questions.

CSSFrancis commented on doc/conf.py:
```diff
@@ -27,21 +27,33 @@
+    "sphinx.ext.inheritance_diagram",
+    "sphinx.ext.intersphinx",
     "sphinx.ext.napoleon",
-    "sphinx.ext.linkcode",
```
I think that was a mistake.... adding that back in

hakonanes replied:
I'm happy you did (: I also prefer linking directly to a more dynamic view of the source on GitHub instead of the static view as part of the docs.

CSSFrancis commented on doc/conf.py:
```diff
@@ -27,21 +27,33 @@
 
 # Add any Sphinx extension module names here, as strings. They can be extensions
 # coming with Sphinx (named 'sphinx.ext.*') or your custom ones.
 extensions = [
```
Ooh good suggestion!

CSSFrancis commented on doc/user_guide/distributed_big_data.rst:
```diff
@@ -0,0 +1,75 @@
+.. code-block:: python
+    >>> haadf = hs.load("haadf.zspy") # load the HAAADF dataset
+    >>> s.naviagtor = haadf # set the navigator to the HAAADF dataset
+    >>> s.plot() # happens instantaneously
```
@hakonanes The idea is that if you save the data the navigator is saved as well (Under .metedata._Hyperspy.navigator). I like to do this mostly because I __hate__ when I accidently call `s.plot()` without a navigator object and it tries to compute the navigator for some very large dataset.

CSSFrancis commented on doc/user_guide/installing.rst:
```diff
@@ -0,0 +1,60 @@
+which is a pre-configured python environment that includes pyxem and all its dependencies.
+
+It includes many other packages that are useful for electron microscopy data analysis, such
+as py4dSTEM, AbTEM, and hyperspy. It also includes the Jupyter notebook and Jupyter lab pre-configured
```
The capitalization drives me nuts some times :)

CSSFrancis commented:
@hakonanes Thanks for the suggestions!  I think I've made most of the needed changes.  I'll probably find a couple more mistakes eventually but for now I think it is good.

ericpre reviewed (commented):
This is very nice! :)

Regarding the tutorials pulled from pyxem-demos, I am wondering if it would be better to use git submodules (https://github.blog/2016-02-01-working-with-submodules or https://gist.github.com/gitaarik/8735255)

The thumbnail of the some of the tutorials, is it something that needs to be fixed here or in the pyxem-demos repository?
![image](https://github.com/pyxem/pyxem/assets/11851990/79f05e76-0c5d-4ccd-b331-8d8404bc39ab)

One thing that would be very useful is to document in the developer guide the workflow around the tutorials (where it lives, how it is build, etc.)

ericpre commented on doc/intro.rst:
```diff
@@ -0,0 +1,70 @@
+pyxem's Documentation
+=====================
+
+Pyxem is an open-source project working to bring together tools for the analysis of
+data from pixelated electron detectors. Based on the HyperSpy project, pyxem has an
+active community of users and developers who are working to develop new tools for
+4-D, 5-D STEM and other pixelated electron detector data.
```
```suggestion
Pyxem is an open-source project working to bring together tools for the analysis of electron diffraction data, in
particular data acquired from pixelated electron detectors. With its roots in the HyperSpy community, pyxem has an
active community of users and developers who are working to provide advanced and state-of-the art tools for the
analysis of 4-D, 5-D STEM and other pixelated electron detector data.
```

CSSFrancis commented:
> This is very nice! :)

@ericpre thanks for looking at this!

> Regarding the tutorials pulled from pyxem-demos, I am wondering if it would be better to use git submodules (https://github.blog/2016-02-01-working-with-submodules or https://gist.github.com/gitaarik/8735255)

That's a good idea!  A couple of questions, I think when I previously tired symlinking the dataset it just checks out some commit, this would as well? Correct?  So we would need to run `git submodule update` in the CI?

> The thumbnail of the some of the tutorials, is it something that needs to be fixed here or in the pyxem-demos repository? ![image](https://user-images.githubusercontent.com/11851990/277104337-79f05e76-0c5d-4ccd-b331-8d8404bc39ab.png)

The thumbnails in the repo don't properly load because the notebooks aren't populated.  The reason for that is originally because we didn't want the Jupyter notebooks to blow up the lines of code changed but I think that doesn't matter too much.  The Read the Docs CI doesn't have enough memory to run the Jupyter Notebooks to auto populate so I haven't figured out a good solution.  

> One thing that would be very useful is to document in the developer guide the workflow around the tutorials (where it lives, how it is build, etc.)

That would be good to add!

ericpre commented:
> That's a good idea! A couple of questions, I think when I previously tired symlinking the dataset it just checks out some commit, this would as well? Correct? So we would need to run git submodule update in the CI?

For CI, it seems it should be possible to configure it:
- readthedocs, : https://docs.readthedocs.io/en/stable/config-file/v2.html#submodules
- GitHub actions: https://github.com/actions/checkout?tab=readme-ov-file#usage

I have never used git submodules so I don't know more than what a quick online search will give you. When reading the script do a manual git clone and I was thinking that maybe it would be better to use a standardised approach.

It may be worth assessing the pros and cons of using git submodules, it seems hat a usual source of confusion/frustation is that you need to update the submodules separately and then commit your changes but it seems that there some settings to facilitate this. In principle, it should be possible to commit the notebook at the same time as changing code in pyxem but it may need a good understanding of git submodules and could be useful?
The submodule may need to be update regularly to be keep up to date in locally when checking out HEAD. On CI, the submodules can be made explicitely.

CSSFrancis commented:
> I have never used git submodules so I don't know more than what a quick online search will give you. When reading the script do a manual git clone and I was thinking that maybe it would be better to use a standardised approach.
> 
> It may be worth assessing the pros and cons of using git submodules, it seems hat a usual source of confusion/frustation is that you need to update the submodules separately and then commit your changes but it seems that there some settings to facilitate this. In principle, it should be possible to commit the notebook at the same time as changing code in pyxem but it may need a good understanding of git submodules and could be useful? The submodule may need to be update regularly to be keep up to date in locally when checking out HEAD. On CI, the submodules can be made explicitely.

The manual updating of the submodules seems to be a bit of a pain.  It seems like Dependabot can update them daily but I'm not sure if you can push changes to both repos at the same time without having some sort of GitHub action which can push directly from the `pyxem-demos` repo to the `pyxem` repo.  That's not a huge deal but is a little bit of a security risk. 

I guess that maybe something like pre-commit could handle it?

hakonanes commented:
nbsphinx allows setting a thumbnail from a file: https://nbsphinx.readthedocs.io/en/0.9.3/gallery/thumbnail-from-conf-py.html. Otherwise nbsphinx must fetch the thumbnail from the last image or a particular notebook cell.

Using a specific file as thumbnail adds to maintenance (will the thumbnail stay relevant?). If we decide to leave notebooks un-completed, one solution is to use the pyxem logo as thumbnail for now.

CSSFrancis commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,10 @@
+    rev: 23.9.1
+    hooks:
+    - id: black
+ci:
```
Sorry that was more a frustration with the current `black` checks  and auto application. I thought they were broken for some reason, but realized there was one file which was failing and not printing out.   I was meaning to move to using pre-commit ci as it is probably the better way to handle applying `black` than what we were previously doing.

CSSFrancis commented:
> nbsphinx allows setting a thumbnail from a file: https://nbsphinx.readthedocs.io/en/0.9.3/gallery/thumbnail-from-conf-py.html. Otherwise nbsphinx must fetch the thumbnail from the last image or a particular notebook cell.
> 
> Using a specific file as thumbnail adds to maintenance (will the thumbnail stay relevant?). If we decide to leave notebooks un-completed, one solution is to use the pyxem logo as thumbnail for now.

I was just going to run all of the notebooks.  I __like__ having them complete on GitHub as I think it gives a better idea of what things should look like.   I also previously went through and set the thumbnail image so that should be good!

CSSFrancis commented:
pre-commit.ci autofix

CSSFrancis commented on doc/check-demos.sh:
```diff
@@ -0,0 +1,11 @@
+#! /bin/bash
```
Yea it probably should be.  I was still trying to figure out what the best way to handle that is though...

ericpre reviewed (approved):
There are a lots when warnings, but I guess, these are not new and would be better to be fixed in a separate PR.

CSSFrancis commented:
> There are a lots when warnings, but I guess, these are not new and would be better to be fixed in a separate PR.

I'll do a follow up for the warnings. Merging...
